cpp/options.cpp

```diff
@@ -0,0 +1,308 @@
+
+
+const OptionValue &OptionValue::operator =(const OptionValue &other) {
+	if (this->type == other.type) {
+		this->set(other);
+	}
+	return *this;
+}
+
+void OptionValue::set(const OptionValue &other) {
+	switch (this->type) {
+
+	case option_type::bool_type:
+		this->set_value<bool>(other);
+		break;
+
+	case option_type::int_type:
+		this->set_value<int>(other);
+		break;
+
+	case option_type::double_type:
+		this->set_value<double>(other);
+		break;
+
+	case option_type::string_type:
+		this->set_value<std::string>(other);
+		break;
+
+	case option_type::list_type:
+		this->set_value<option_list>(other);
+		break;
+
+	}
+}
+
+
+std::string OptionValue::str_value() const {
+	switch (this->type) {
+
+	case option_type::bool_type:
+		return this->value<bool>()? "true" : "false";
+
+	case option_type::int_type:
+		return std::to_string(this->value<int>());
+
+	case option_type::double_type:
+		return std::to_string(this->value<double>());
+
+	case option_type::string_type:
+		return this->value<std::string>();
+
+	case option_type::list_type: {
+		std::string result = "[ ";
+		for (auto &i : this->value<option_list>()) {
+			result += i.str_value() + " ";
+		}
+		result += "]";
+		return result;
+	}
+
+	}
+	return "";
+}
+
+OptionValue parse(option_type t, std::string s) {
+	switch(t) {
+	case options::option_type::bool_type:
+		return options::OptionValue(s == "true");
+	case option_type::int_type:
+		return options::OptionValue(stoi(s));
+	case option_type::double_type:
+		return options::OptionValue(stod(s));
+	case option_type::string_type:
+		return options::OptionValue(s);
+	case option_type::list_type:
+		// TODO:
+		return options::OptionValue(false);
+	}
+	return options::OptionValue(false);
+}
+
+
+
+OptionAction::OptionAction(const std::string &name, const opt_func_t f)
+	:
+	name{name},
+	function{f} {
+}
+
+
+bool OptionAction::do_action() {
+	this->function();
+	return true;
+}
+
+
+OptionNode::OptionNode(const std::string &panel_name)
+	:
+	name{panel_name},
+	parent{nullptr} {
+}
+
+
+OptionNode::~OptionNode() {
+	if (this->parent) {
+		this->parent->remove_panel(this);
+	}
+}
+
+
+std::vector<std::string> OptionNode::list_options(bool recurse, std::string indent) {
+	std::vector<std::string> result;
+	result.push_back(indent + "node " + this->name + " {");
+	std::string inner_indent = indent + "\t";
+	for (auto &p : bools) {
+		result.push_back(inner_indent + "bool " + p.first + " = " + (*p.second ? "true" : "false"));
+	}
+	for (auto &p : ints) {
+		result.push_back(inner_indent + "int " + p.first + " = " + std::to_string(*p.second));
+	}
+	for (auto &v : varmap) {
+		result.push_back(inner_indent + "var " + v.first + " = " + v.second.str_value());
+	}
+	for (auto &f : actions) {
+		result.push_back(inner_indent + "func " + f.first + "()");
+	}
+	for (auto &p : panels) {
+		for (auto &line : p.second->list_options(recurse, inner_indent)) {
+			result.push_back(line);
+		}
+	}
+	result.push_back(indent + "}");
+	return result;
```

Comment:
for all these stack-constructed vectors we should use `std::move` for returning them.